# Regression Exercises I
Let's review the steps we take at the beginning of each new module.

1. Create a new repository named ```regression-exercises``` in your ```GitHub```; all of your Regression work will be housed here.
2. Clone this repository within your local ```codeup-data-science``` directory.
3. Create a ```.gitignore``` and make sure your list of 'files to ignore' includes your ```env.py``` file.
4. Ceate a ```README.md``` file that outlines the contents and purpose of your repository.
5. ```Add```, ```commit```, and ```push``` these two files.
6. Now you can add your ```env.py``` file to this repository to access the ```Codeup database server```.
7. For these exercises, you will create ```wrangle.ipynb``` and ```wrangle.py``` files to hold necessary functions.
8. As always, ```add```, ```commit```, and ```push``` your work often.

# Regression Exercises II
Let's set up an example scenario as perspective for our regression exercises using the ```Zillow dataset```.

As a Codeup data science graduate, you want to show off your skills to the Zillow data science team in hopes of getting an interview for a position you saw pop up on [LinkedIn](www.LinkedIn.com). You thought it might look impressive to build an end-to-end project in which you use some of their Kaggle data to predict property values using some of their available features; who knows, you might even do some feature engineering to blow them away. 

##### Your goal is to predict the values of single unit properties using the obervations from 2017.

In these exercises, you will complete the first step toward the above goal: ```acquire``` and ```prepare``` the necessary ```Zillow data``` from the ```zillow database``` in the ```Codeup database server```.

In [1]:
import numpy as np
import pandas as pd

import QMCBT_wrangle as w
import QMCBT_explore_evaluate as ee
from env import user, password, host

# 1. ```Acquire``` 
* bedroomcnt 
* bathroomcnt 
* calculatedfinishedsquarefeet
* taxvaluedollarcnt 
* yearbuilt 
* taxamount  
* fips 


from the ```zillow database``` for all '```Single Family Residential```' properties.

In [2]:
db = 'zillow'
connection = f'mysql+pymysql://{user}:{password}@{host}/{db}'
query = """SELECT propertylandusetypeid, propertylandusedesc, bedroomcnt, bathroomcnt, 
calculatedfinishedsquarefeet, taxvaluedollarcnt, yearbuilt, taxamount, fips 
FROM properties_2017 
LEFT JOIN propertylandusetype 
USING (propertylandusetypeid) 
WHERE propertylandusetypeid = 261"""

In [3]:
df = pd.read_sql(query, connection)

# 2. Using your acquired Zillow data, walk through the summarization and cleaning steps in your ```wrangle.ipynb``` file like we did above. You may handle the missing values however you feel is appropriate and meaningful; remember to document your process and decisions using markdown and code commenting where helpful.

<div class="alert alert-info">

## Preparation
* ### Check is blank
    * #### Remove Whitespace
    * ```df = df.replace(r'^\s*$', np.NaN, regex=True)```


* ### Check is Null
    * #### Remove NULL/NaN
    * ```df = df.dropna()```
    
* ### Check dtype
    * #### Convert dtypes
    * ```df = df.convert_dtypes(infer_objects=False)```

* ### Remove Outliers
    * #### remove homes with no bedrooms or bathrooms
    * ```df = df[df.bedroomcnt > 0]```
    * ```df = df[df.bathroomcnt > 0]```

    * #### remove homes with more than 8 bedrooms or bathrooms
    * ```df = df[df.bedroomcnt <= 8]```
    * ```df = df[df.bathroomcnt <= 8]```

    * #### remove homes with tax value of less than 50k or more than 2mil
    * ```df = df[df.taxvaluedollarcnt > 50_000]```
    * ```df = df[df.taxvaluedollarcnt < 2_000_000]```

    * #### remove sqft less than 400 or more than 10,000```
    * ```df = df[df.calculatedfinishedsquarefeet < 10_000]```
    * ```df = df[df.calculatedfinishedsquarefeet > 400]```
    
    * #### remove tax percent of less than 1% and more than 100%
    * ```df = df[df.taxpercent > .0099]```
    * ```df = df[df.taxpercent < 1]```


* ### Feature Engineer
    * #### Create a feature to replace yearbuilt that shows the age of the home in 2017 when data was collected
    * ```df['age'] = 2017 - df.yearbuilt```

    * #### Create a feature to show tax percentage of value
    * ```df['taxpercent'] = round((df.taxamount / df.taxvaluedollarcnt), 4)```

    * #### Create a feature to show Bathroom to Bedroom ratio
    * ``` CODE ```


* ### Maintain Columns
    * #### Drop Columns
    * ```df = df.drop(columns=['propertylandusetypeid', 'propertylandusedesc'])```
    * #### Organize Columns
    * ``` CODE ```
    * #### Rename Columns
    * ``` CODE ```
    

</div>

In [4]:
df.T

0        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                               27516.0   
yearbuilt                                           NaN   
taxamount                                           NaN   
fips                                             6037.0   

                                                1        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                                  10.0   
yearbuilt                                           NaN   
taxamount                                           NaN   
fips                                             6037.0   

                                                2        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                                  10.0   
yearbuilt                                           NaN   
taxamount                                           NaN   
fips                                             6037.0   

                                                3        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                                2108.0   
yearbuilt                                           NaN   
taxamount                                        174.21   
fips                                             6037.0   

                                                4        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          4.0   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                     3633.0   
taxvaluedollarcnt                              296425.0   
yearbuilt                                        2005.0   
taxamount                                       6941.39   
fips                                             6037.0   

                                                5        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                                 124.0   
yearbuilt                                           NaN   
taxamount                                           NaN   
fips                                             6037.0   

                                                6        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          3.0   
bathroomcnt                                         4.0   
calculatedfinishedsquarefeet                     1620.0   
taxvaluedollarcnt                              847770.0   
yearbuilt                                        2011.0   
taxamount                     

In [5]:
df.shape

(2152863, 9)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2152863 entries, 0 to 2152862
Data columns (total 9 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   propertylandusetypeid         float64
 1   propertylandusedesc           object 
 2   bedroomcnt                    float64
 3   bathroomcnt                   float64
 4   calculatedfinishedsquarefeet  float64
 5   taxvaluedollarcnt             float64
 6   yearbuilt                     float64
 7   taxamount                     float64
 8   fips                          float64
dtypes: float64(8), object(1)
memory usage: 147.8+ MB


In [7]:
df.describe().T

count           mean            std  \
propertylandusetypeid         2152863.0     261.000000       0.000000   
bedroomcnt                    2152852.0       3.287196       0.954754   
bathroomcnt                   2152852.0       2.230688       0.999280   
calculatedfinishedsquarefeet  2144379.0    1862.855178    1222.125124   
taxvaluedollarcnt             2152370.0  461896.237963  699676.049600   
yearbuilt                     2143526.0    1960.949681      22.162196   
taxamount                     2148421.0    5634.865978    8178.910249   
fips                          2152863.0    6048.377335      20.433292   

                                  min        25%        50%        75%  \
propertylandusetypeid          261.00     261.00     261.00     261.00   
bedroomcnt                       0.00       3.00       3.00       4.00   
bathroomcnt                      0.00       2.00       2.00       3.00   
calculatedfinishedsquarefeet     1.00    1257.00    1623.00    2208.00   
taxvaluedollarcnt                1.00  188170.25  327671.00  534527.00   
yearbuilt                     1801.00    1949.00    1958.00    1976.00   
taxamount                        1.85    2534.98    4108.95    6414.32   
fips                          6037.00    6037.00    6037.00    6059.00   

                                      max  
propertylandusetypeid              261.00  
bedroomcnt                          25.00  
bathroomcnt                         32.00  
calculatedfinishedsquarefeet    952576.00  
taxvaluedollarcnt             98428909.00  
yearbuilt                         2016.00  
taxamount                      1337755.86  
fips                              6111.00

## Check isblank

In [8]:
# Return (row count)
row_count = df.shape[0]
row_count

2152863

In [9]:
# creates list of columns
column_list = df.columns
column_list

Index(['propertylandusetypeid', 'propertylandusedesc', 'bedroomcnt',
       'bathroomcnt', 'calculatedfinishedsquarefeet', 'taxvaluedollarcnt',
       'yearbuilt', 'taxamount', 'fips'],
      dtype='object')

In [10]:
# returns count of all rows from each column that has values
# essentially skipping count on any row that has a column with whitespace
row_value_count = df[column_list].value_counts().sum()
row_value_count

2140235

In [11]:
# subtract value count from row count to get count of rows with whitespace
whitespace_count = row_count - row_value_count
whitespace_count

12628

In [12]:
# Clean whitespace using R trick shared by Codeup Instructor Madeleine Capper

# teensy tiny regex mini lesson for very smols
# ^ : "starts with"
# \s : "any type of whitespace"
# * : " zero or more times"
# $ : "ends with"

# '^\s*$' : something that starts with any whitespace
          # character for zero or more times until the end

# let's change the whitespace into a null value,
# since thats effectively what it is
# Note: regex is not the only option/way to do this
# but its an awfully convenient one that I want y'all to see

# wrapping it in df = without specifying a specific column will 
    # remove ALL whitespace and update the df
    
df = df.replace(r'^\s*$', np.NaN, regex=True)

## Check isnull
* Create Function to automate this to one level of NaN check or figure out how to loop it
    * found single line of code in wrangle file created by Codeup instructor Adam King
    * ```df.dropna().shape[0] / df.shape[0]```
    * turned this into a function ```w.null_stats(df)```
* The next 29 lines of code show my journey to accomplish what that single line of code does

In [13]:
df.isnull().sum()

propertylandusetypeid              0
propertylandusedesc                0
bedroomcnt                        11
bathroomcnt                       11
calculatedfinishedsquarefeet    8484
taxvaluedollarcnt                493
yearbuilt                       9337
taxamount                       4442
fips                               0
dtype: int64

In [14]:
# assuming that even if all NaNs lived in separate rows
# if we divide the number of NaNs by the number of rows
# we will get the overall percentage of NaN rows for potential deletion
NaN_count = (11+11+8484+493+9337+4442)
NaN_count, NaN_count/row_count

(22778, 0.01058032954256727)

In [15]:
# let's take our largest NaN count column and filter by NaN
isnull_df = df[df.yearbuilt.isnull()==True]

In [16]:
# check our work
isnull_df.T

0        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                               27516.0   
yearbuilt                                           NaN   
taxamount                                           NaN   
fips                                             6037.0   

                                                1        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                                  10.0   
yearbuilt                                           NaN   
taxamount                                           NaN   
fips                                             6037.0   

                                                2        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                                  10.0   
yearbuilt                                           NaN   
taxamount                                           NaN   
fips                                             6037.0   

                                                3        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                                2108.0   
yearbuilt                                           NaN   
taxamount                                        174.21   
fips                                             6037.0   

                                                5        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                                 124.0   
yearbuilt                                           NaN   
taxamount                                           NaN   
fips                                             6037.0   

                                                8        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                             6730242.0   
yearbuilt                                           NaN   
taxamount                                      80348.13   
fips                                             6037.0   

                                                9        \
propertylandusetypeid                             261.0   
propertylandusedesc           Single Family Residential   
bedroomcnt                                          0.0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        NaN   
taxvaluedollarcnt                               15532.0   
yearbuilt                                           NaN   
taxamount                     

In [17]:
# now let's sum NaNs of all columns within just that filter
isnull_df.isnull().sum()

propertylandusetypeid              0
propertylandusedesc                0
bedroomcnt                         4
bathroomcnt                        4
calculatedfinishedsquarefeet    7877
taxvaluedollarcnt                375
yearbuilt                       9337
taxamount                       1734
fips                               0
dtype: int64

In [18]:
isnull_df.isnull().sum().max()

9337

In [19]:
# Returns (isnull row count)
# This will show us how many records would be deleted
isnull_row_count = isnull_df.shape[0]
isnull_row_count

9337

In [20]:
# now we can easily compare our first NaN_count against our filtered_NaN_count
# this show us that there are only 3,447 NaNs that do not coexist within the largest NaN column
filtered_NaN_count = (4+4+7877+375+9337+1734)
filtered_NaN_count, NaN_count - filtered_NaN_count

(19331, 3447)

In [21]:
# 9337 + (22778 - 19331)
# This equals the potential row count of NaNs to be deleted
NaN_delete_row_count = isnull_row_count + (NaN_count - filtered_NaN_count)
NaN_delete_row_count

12784

In [22]:
# assuming that even if all additional NaNs lived in separate rows
# we can now more accurately see the percentage of filtered NaNs
    # plus additional NaNs against the overall database
NaN_delete_row_count / row_count 

0.005938139119860391

<div class="alert alert-info">
    
As you can see deleting <b>less than*</b> 12,784 records or rows will result in a less than 1% data loss.  
    
<b>less than*</b> - because we assume all additional NaNs (3,447) are individual rows when in reality many probably share the same rows.
    
### Therefore, it is reasonably safe to remove all NaN records from the DataFrame.</div>

In [24]:
# Check the percentage of dropping NaN
df.dropna().shape[0], df.shape[0], 1 - (df.dropna().shape[0] / df.shape[0])

(2140235, 2152863, 0.0058656774722776195)

In [25]:
w.null_stats(df)

COUNT OF NULL/NaN PER COLUMN:
propertylandusetypeid              0
propertylandusedesc                0
bedroomcnt                        11
bathroomcnt                       11
calculatedfinishedsquarefeet    8484
taxvaluedollarcnt                493
yearbuilt                       9337
taxamount                       4442
fips                               0
dtype: int64

     DataFrame Row Count: 2152863
      NULL/NaN Row Count: 2140235
  DataFrame Percent kept: 0.9941
NULL/NaN Percent dropped: 0.0059


In [26]:
# Now, let's remove all of the NaN's 
    # since we know they are insignificant in size and percentage
df = df.dropna()

In [27]:
# Check our work
df.isnull().sum()

propertylandusetypeid           0
propertylandusedesc             0
bedroomcnt                      0
bathroomcnt                     0
calculatedfinishedsquarefeet    0
taxvaluedollarcnt               0
yearbuilt                       0
taxamount                       0
fips                            0
dtype: int64

## Clean dtypes

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2140235 entries, 4 to 2152862
Data columns (total 9 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   propertylandusetypeid         float64
 1   propertylandusedesc           object 
 2   bedroomcnt                    float64
 3   bathroomcnt                   float64
 4   calculatedfinishedsquarefeet  float64
 5   taxvaluedollarcnt             float64
 6   yearbuilt                     float64
 7   taxamount                     float64
 8   fips                          float64
dtypes: float64(8), object(1)
memory usage: 163.3+ MB


In [31]:
# Function will show best dtype based on values (ignore objects)
# be sure to double check that the computer got this right before converting
df.convert_dtypes(infer_objects=False).dtypes

propertylandusetypeid             Int64
propertylandusedesc              string
bedroomcnt                        Int64
bathroomcnt                     Float64
calculatedfinishedsquarefeet      Int64
taxvaluedollarcnt                 Int64
yearbuilt                         Int64
taxamount                       Float64
fips                              Int64
dtype: object

In [32]:
# no changes were made by convert_dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2140235 entries, 4 to 2152862
Data columns (total 9 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   propertylandusetypeid         float64
 1   propertylandusedesc           object 
 2   bedroomcnt                    float64
 3   bathroomcnt                   float64
 4   calculatedfinishedsquarefeet  float64
 5   taxvaluedollarcnt             float64
 6   yearbuilt                     float64
 7   taxamount                     float64
 8   fips                          float64
dtypes: float64(8), object(1)
memory usage: 227.8+ MB


In [33]:
# make actual conversion
df = df.convert_dtypes(infer_objects=False)

In [34]:
# Check work
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2140235 entries, 4 to 2152862
Data columns (total 9 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   propertylandusetypeid         Int64  
 1   propertylandusedesc           string 
 2   bedroomcnt                    Int64  
 3   bathroomcnt                   Float64
 4   calculatedfinishedsquarefeet  Int64  
 5   taxvaluedollarcnt             Int64  
 6   yearbuilt                     Int64  
 7   taxamount                     Float64
 8   fips                          Int64  
dtypes: Float64(2), Int64(6), string(1)
memory usage: 244.1 MB


In [35]:
# Check work
df.T

4        \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            4   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       3633   
taxvaluedollarcnt                                296425   
yearbuilt                                          2005   
taxamount                                       6941.39   
fips                                               6037   

                                                6        \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         4.0   
calculatedfinishedsquarefeet                       1620   
taxvaluedollarcnt                                847770   
yearbuilt                                          2011   
taxamount                                      10244.94   
fips                                               6037   

                                                7        \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       2077   
taxvaluedollarcnt                                646760   
yearbuilt                                          1926   
taxamount                                       7924.68   
fips                                               6037   

                                                11       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                       1200   
taxvaluedollarcnt                                  5328   
yearbuilt                                          1972   
taxamount                                          91.6   
fips                                               6037   

                                                14       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        171   
taxvaluedollarcnt                                  6920   
yearbuilt                                          1973   
taxamount                                        255.17   
fips                                               6037   

                                                15       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        203   
taxvaluedollarcnt                                 14166   
yearbuilt                                          1960   
taxamount                                        163.79   
fips                                               6037   

                                                18       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         1.0   
calculatedfinishedsquarefeet                       1244   
taxvaluedollarcnt                                169471   
yearbuilt                                          1950   
taxamount                     

### Remove Outliers

In [36]:
# Is this even possible? Are these errors?
# Do we need to drop rows < 300? 600? 
df.loc[df['calculatedfinishedsquarefeet'] == 1].T

58438    \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            2   
bathroomcnt                                         1.0   
calculatedfinishedsquarefeet                          1   
taxvaluedollarcnt                                121376   
yearbuilt                                          1907   
taxamount                                       1996.35   
fips                                               6037   

                                                1046787  \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                          1   
taxvaluedollarcnt                                 28091   
yearbuilt                                          1963   
taxamount                                        439.55   
fips                                               6037   

                                                1276353  \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            1   
bathroomcnt                                         3.0   
calculatedfinishedsquarefeet                          1   
taxvaluedollarcnt                                124906   
yearbuilt                                          1953   
taxamount                                       2020.66   
fips                                               6037   

                                                1359288  \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            2   
bathroomcnt                                         1.0   
calculatedfinishedsquarefeet                          1   
taxvaluedollarcnt                                147577   
yearbuilt                                          1991   
taxamount                                        1855.4   
fips                                               6037   

                                                1895866  \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            5   
bathroomcnt                                         5.0   
calculatedfinishedsquarefeet                          1   
taxvaluedollarcnt                                563977   
yearbuilt                                          1997   
taxamount                                       6808.84   
fips                                               6037   

                                                2017745  
propertylandusetypeid                               261  
propertylandusedesc           Single Family Residential  
bedroomcnt                                            3  
bathroomcnt                                         1.0  
calculatedfinishedsquarefeet                          1  
taxvaluedollarcnt                                 31800  
yearbuilt                                          1900  
taxamount                                        870.36  
fips                                               6037

In [51]:
# why do these homes have no bedrooms or bathrooms?
df.loc[df['bedroomcnt'] == 0].T 

11       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                       1200   
taxvaluedollarcnt                                  5328   
yearbuilt                                          1972   
taxamount                                          91.6   
fips                                               6037   

                                                14       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        171   
taxvaluedollarcnt                                  6920   
yearbuilt                                          1973   
taxamount                                        255.17   
fips                                               6037   

                                                15       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        203   
taxvaluedollarcnt                                 14166   
yearbuilt                                          1960   
taxamount                                        163.79   
fips                                               6037   

                                                866      \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        220   
taxvaluedollarcnt                                 13840   
yearbuilt                                          1988   
taxamount                                         327.8   
fips                                               6037   

                                                2306     \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                       1776   
taxvaluedollarcnt                                 79574   
yearbuilt                                          1989   
taxamount                                        1137.9   
fips                                               6037   

                                                2313     \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        352   
taxvaluedollarcnt                                 25464   
yearbuilt                                          1948   
taxamount                                        415.92   
fips                                               6037   

                                                2676     \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                       1057   
taxvaluedollarcnt                                344835   
yearbuilt                                          1942   
taxamount                     

In [52]:
df.loc[df['bathroomcnt'] == 0].T

11       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                       1200   
taxvaluedollarcnt                                  5328   
yearbuilt                                          1972   
taxamount                                          91.6   
fips                                               6037   

                                                14       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        171   
taxvaluedollarcnt                                  6920   
yearbuilt                                          1973   
taxamount                                        255.17   
fips                                               6037   

                                                15       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        203   
taxvaluedollarcnt                                 14166   
yearbuilt                                          1960   
taxamount                                        163.79   
fips                                               6037   

                                                398      \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                       2239   
taxvaluedollarcnt                                396608   
yearbuilt                                          1988   
taxamount                                       4404.22   
fips                                               6111   

                                                753      \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                       2421   
taxvaluedollarcnt                               1111120   
yearbuilt                                          1976   
taxamount                                      12394.16   
fips                                               6111   

                                                866      \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            0   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        220   
taxvaluedollarcnt                                 13840   
yearbuilt                                          1988   
taxamount                                         327.8   
fips                                               6037   

                                                1546     \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            1   
bathroomcnt                                         0.0   
calculatedfinishedsquarefeet                        400   
taxvaluedollarcnt                                 47682   
yearbuilt                                          1913   
taxamount                     

In [72]:
# filter down outliers to more accurately align with realistic expectations of a Single Family Residence

# Set no_outliers equal to df
no_outliers = df

# Keep all homes that have > 0 and <= 8 Beds and Baths
no_outliers = no_outliers[no_outliers.bedroomcnt > 0]
no_outliers = no_outliers[no_outliers.bathroomcnt > 0]
no_outliers = no_outliers[no_outliers.bedroomcnt <= 8]
no_outliers = no_outliers[no_outliers.bathroomcnt <= 8]

# Keep all homes that have tax value > 30 thousand and <= 2 million
no_outliers = no_outliers[no_outliers.taxvaluedollarcnt >= 50_000]
no_outliers = no_outliers[no_outliers.taxvaluedollarcnt <= 2_000_000]

# Keep all homes that have sqft > 4 hundred and < 10 thousand
no_outliers = no_outliers[no_outliers.calculatedfinishedsquarefeet > 400]
no_outliers = no_outliers[no_outliers.calculatedfinishedsquarefeet < 10_000]

no_outliers.T

4        \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            4   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       3633   
taxvaluedollarcnt                                296425   
yearbuilt                                          2005   
taxamount                                       6941.39   
fips                                               6037   
age                                                  12   
taxpercent                                       0.0234   

                                                6        \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         4.0   
calculatedfinishedsquarefeet                       1620   
taxvaluedollarcnt                                847770   
yearbuilt                                          2011   
taxamount                                      10244.94   
fips                                               6037   
age                                                   6   
taxpercent                                       0.0121   

                                                7        \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       2077   
taxvaluedollarcnt                                646760   
yearbuilt                                          1926   
taxamount                                       7924.68   
fips                                               6037   
age                                                  91   
taxpercent                                       0.0123   

                                                18       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         1.0   
calculatedfinishedsquarefeet                       1244   
taxvaluedollarcnt                                169471   
yearbuilt                                          1950   
taxamount                                       2532.88   
fips                                               6037   
age                                                  67   
taxpercent                                       0.0149   

                                                19       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       1300   
taxvaluedollarcnt                                233266   
yearbuilt                                          1950   
taxamount                                       3110.99   
fips                                               6037   
age                                                  67   
taxpercent                                       0.0133   

                                                20       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       1222   
taxvaluedollarcnt                                290492   
yearbuilt                                          1951   
taxamount                      

In [73]:
# Use code from NULL/NaN status function to check percentage of files to be deleted
print(f'     DataFrame Row Count: {df.shape[0]}')
print(f'   no_outliers Row Count: {no_outliers.shape[0]}')
print(f'  DataFrame Percent kept: {round(no_outliers.shape[0] / df.shape[0], 4)}')
print(f'Outliers Percent dropped: {round(1 - (no_outliers.shape[0] / df.shape[0]), 4)}')

     DataFrame Row Count: 2086058
   no_outliers Row Count: 2041634
  DataFrame Percent kept: 0.9787
Outliers Percent dropped: 0.0213


In [63]:
df = no_outliers

In [64]:
df.T

4        \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            4   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       3633   
taxvaluedollarcnt                                296425   
yearbuilt                                          2005   
taxamount                                       6941.39   
fips                                               6037   

                                                6        \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         4.0   
calculatedfinishedsquarefeet                       1620   
taxvaluedollarcnt                                847770   
yearbuilt                                          2011   
taxamount                                      10244.94   
fips                                               6037   

                                                7        \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       2077   
taxvaluedollarcnt                                646760   
yearbuilt                                          1926   
taxamount                                       7924.68   
fips                                               6037   

                                                18       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         1.0   
calculatedfinishedsquarefeet                       1244   
taxvaluedollarcnt                                169471   
yearbuilt                                          1950   
taxamount                                       2532.88   
fips                                               6037   

                                                19       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       1300   
taxvaluedollarcnt                                233266   
yearbuilt                                          1950   
taxamount                                       3110.99   
fips                                               6037   

                                                20       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       1222   
taxvaluedollarcnt                                290492   
yearbuilt                                          1951   
taxamount                                       3870.25   
fips                                               6037   

                                                21       \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            4   
bathroomcnt                                         4.0   
calculatedfinishedsquarefeet                       4144   
taxvaluedollarcnt                               1303522   
yearbuilt                                          2016   
taxamount                     

In [66]:
df.describe().T

count           mean            std  \
propertylandusetypeid         2086058.0     261.000000       0.000000   
bedroomcnt                    2086058.0       3.291885       0.893533   
bathroomcnt                   2086058.0       2.200883       0.889556   
calculatedfinishedsquarefeet  2086058.0    1810.227657     799.963383   
taxvaluedollarcnt             2086058.0  402250.352206  315493.357614   
yearbuilt                     2086058.0    1960.846105      21.795807   
taxamount                     2086058.0    4951.648862    3662.682626   
fips                          2086058.0    6048.396252      20.421661   

                                   min        25%        50%        75%  \
propertylandusetypeid           261.00     261.00     261.00     261.00   
bedroomcnt                        1.00       3.00       3.00       4.00   
bathroomcnt                       0.50       2.00       2.00       3.00   
calculatedfinishedsquarefeet    401.00    1257.00    1614.00    2170.00   
taxvaluedollarcnt             30000.00  188830.00  324451.00  519702.50   
yearbuilt                      1801.00    1949.00    1958.00    1975.00   
taxamount                        16.48    2537.39    4066.98    6245.16   
fips                           6037.00    6037.00    6037.00    6059.00   

                                     max  
propertylandusetypeid             261.00  
bedroomcnt                          8.00  
bathroomcnt                         8.00  
calculatedfinishedsquarefeet     9978.00  
taxvaluedollarcnt             2000000.00  
yearbuilt                        2016.00  
taxamount                      100379.34  
fips                             6111.00

In [67]:
df.shape

(2086058, 9)

## Feature Engineer

### Age

In [68]:
# Create a feature to replace yearbuilt that shows the age of the home in 2017 when data was collected
df['age'] = 2017 - df.yearbuilt

### Tax Percentage

In [69]:
# Create a feature to show tax percentage of value
df['taxpercent'] = round((df.taxamount / df.taxvaluedollarcnt), 4)

In [70]:
df.describe().T

count           mean            std  \
propertylandusetypeid         2086058.0     261.000000       0.000000   
bedroomcnt                    2086058.0       3.291885       0.893533   
bathroomcnt                   2086058.0       2.200883       0.889556   
calculatedfinishedsquarefeet  2086058.0    1810.227657     799.963383   
taxvaluedollarcnt             2086058.0  402250.352206  315493.357614   
yearbuilt                     2086058.0    1960.846105      21.795807   
taxamount                     2086058.0    4951.648862    3662.682626   
fips                          2086058.0    6048.396252      20.421661   
age                           2086058.0      56.153895      21.795807   
taxpercent                    2086058.0       0.013182       0.003867   

                                   min          25%          50%          75%  \
propertylandusetypeid           261.00     261.0000     261.0000     261.0000   
bedroomcnt                        1.00       3.0000       3.0000       4.0000   
bathroomcnt                       0.50       2.0000       2.0000       3.0000   
calculatedfinishedsquarefeet    401.00    1257.0000    1614.0000    2170.0000   
taxvaluedollarcnt             30000.00  188830.0000  324451.0000  519702.5000   
yearbuilt                      1801.00    1949.0000    1958.0000    1975.0000   
taxamount                        16.48    2537.3900    4066.9800    6245.1600   
fips                           6037.00    6037.0000    6037.0000    6059.0000   
age                               1.00      42.0000      59.0000      68.0000   
taxpercent                        0.00       0.0118       0.0124       0.0136   

                                       max  
propertylandusetypeid         2.610000e+02  
bedroomcnt                    8.000000e+00  
bathroomcnt                   8.000000e+00  
calculatedfinishedsquarefeet  9.978000e+03  
taxvaluedollarcnt             2.000000e+06  
yearbuilt                     2.016000e+03  
taxamount                     1.003793e+05  
fips                          6.111000e+03  
age                           2.160000e+02  
taxpercent                    4.281000e-01

In [71]:
round(df.taxpercent.describe(), 3)

count    2086058.000
mean           0.013
std            0.004
min            0.000
25%            0.012
50%            0.012
75%            0.014
max            0.428
Name: taxpercent, dtype: float64

In [74]:
del_df = df
del_df = df[df.taxpercent < .0099]
del_df = df[df.taxpercent >= .03]
del_df.T

1403     \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       1710   
taxvaluedollarcnt                                131886   
yearbuilt                                          1952   
taxamount                                      12797.47   
fips                                               6037   
age                                                  65   
taxpercent                                        0.097   

                                                1574     \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         1.0   
calculatedfinishedsquarefeet                       1123   
taxvaluedollarcnt                                 39920   
yearbuilt                                          1954   
taxamount                                       2957.35   
fips                                               6037   
age                                                  63   
taxpercent                                       0.0741   

                                                1696     \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            2   
bathroomcnt                                         1.0   
calculatedfinishedsquarefeet                       1313   
taxvaluedollarcnt                                206438   
yearbuilt                                          1945   
taxamount                                       6514.88   
fips                                               6037   
age                                                  72   
taxpercent                                       0.0316   

                                                1983     \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       1208   
taxvaluedollarcnt                                 87536   
yearbuilt                                          1958   
taxamount                                       2908.37   
fips                                               6037   
age                                                  59   
taxpercent                                       0.0332   

                                                2064     \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            2   
bathroomcnt                                         1.0   
calculatedfinishedsquarefeet                        635   
taxvaluedollarcnt                                 33705   
yearbuilt                                          1949   
taxamount                                       1101.12   
fips                                               6037   
age                                                  68   
taxpercent                                       0.0327   

                                                2183     \
propertylandusetypeid                               261   
propertylandusedesc           Single Family Residential   
bedroomcnt                                            3   
bathroomcnt                                         2.0   
calculatedfinishedsquarefeet                       1022   
taxvaluedollarcnt                                 46769   
yearbuilt                                          1957   
taxamount                      

In [75]:
print(f'   DataFrame Row Count: {df.shape[0]}')
print(f'      del_df Row Count: {del_df.shape[0]}')
print(f'DataFrame Percent kept: {round(1 - (del_df.shape[0] / df.shape[0]), 4)}')
print(f'del_df Percent dropped: {round(del_df.shape[0] / df.shape[0], 4)}')

   DataFrame Row Count: 2086058
      del_df Row Count: 7872
DataFrame Percent kept: 0.9962
del_df Percent dropped: 0.0038


In [76]:
del_df.describe().T

count           mean            std       min  \
propertylandusetypeid         7872.0     261.000000       0.000000    261.00   
bedroomcnt                    7872.0       3.084223       0.848831      1.00   
bathroomcnt                   7872.0       1.821900       0.755810      1.00   
calculatedfinishedsquarefeet  7872.0    1478.066311     592.019141    412.00   
taxvaluedollarcnt             7872.0  135022.428735  130248.499085  30053.00   
yearbuilt                     7872.0    1954.283155      20.542248   1880.00   
taxamount                     7872.0    5941.700849    5450.340750    908.65   
fips                          7872.0    6042.456301      15.707371   6037.00   
age                           7872.0      62.716845      20.542248      1.00   
taxpercent                    7872.0       0.049435       0.031369      0.03   

                                     25%        50%          75%           max  
propertylandusetypeid           261.0000    261.000     261.0000  2.610000e+02  
bedroomcnt                        3.0000      3.000       4.0000  8.000000e+00  
bathroomcnt                       1.0000      2.000       2.0000  6.000000e+00  
calculatedfinishedsquarefeet   1095.7500   1330.000    1704.2500  7.682000e+03  
taxvaluedollarcnt             43441.7500  89708.000  189862.2500  1.930658e+06  
yearbuilt                      1947.0000   1954.000    1962.0000  2.016000e+03  
taxamount                      1916.3250   5004.645    8035.3575  1.003793e+05  
fips                           6037.0000   6037.000    6037.0000  6.111000e+03  
age                              55.0000     63.000      70.0000  1.370000e+02  
taxpercent                        0.0324      0.037       0.0526  4.281000e-01

### Bed to Bath Ratio

In [ ]:
# Create a feature to show ration of Bathrooms to Bedrooms
df['bed_bath_ratio'] = round((df.bedroomcnt / df.bathroomcnt), 4)

In [ ]:
df.T

In [ ]:
df.isnull().sum()

### fips Conversion
* This is technically a backwards engineered feature
* fips is already an engineered feature of combining county and state into one code
* This exercise was just a rabit hole for exercise and experience

In [ ]:
# Print Unique List of fips values
df.fips.unique()

Google Results for Reference  
* 6037,Los Angeles County,CA  
* 6059,Orange County,CA  
* 6111,Ventura County,CA  

In [ ]:
# Found a csv fips master list on github
# import using curl
!curl -O https://raw.githubusercontent.com/kjhealy/fips-codes/master/state_and_county_fips_master.csv

# Check to see if it is there
!ls

In [ ]:
# Read it in as a DataFrame
fips_df = pd.read_csv('state_and_county_fips_master.csv')

In [ ]:
# Look at it
fips_df.head()

In [ ]:
# Display just the fips that exist in our zillow df to ensure they exist
# I could also do this by pulling a list from zillow and using the in function
fips6037 = fips_df[fips_df.fips == 6037]
fips6059 = fips_df[fips_df.fips == 6059]
fips6111 = fips_df[fips_df.fips == 6111]
zillow_fips_df = pd.concat([fips6037, fips6059, fips6111], ignore_index=True)
zillow_fips_df

In [ ]:
# left merge to join the name and state to the original df
left_merged_fips_df = pd.merge(df, fips_df, how="left", on=["fips"])

In [ ]:
# Check that it worked
left_merged_fips_df.T

In [ ]:
# Rewrite the df
df = left_merged_fips_df

In [ ]:
df.info()

In [ ]:
df.describe(include='all').T

In [ ]:
# filter down outliers to more accurately align with realistic expectations of a Single Family Residence
    
# remove homes with no bedrooms or bathrooms
df = df[df.bedroomcnt > 0]
df = df[df.bathroomcnt > 0]

# remove homes with more than 8 bedrooms or bathrooms
df = df[df.bedroomcnt <= 8]
df = df[df.bathroomcnt <= 8]

# remove homes with tax value of less than $50k or more than $2 million
df = df[df.taxvaluedollarcnt > 50000]
df = df[df.taxvaluedollarcnt < 2_000_000]

# remove sqft less than 400 or more than 10,000
df = df[df.calculatedfinishedsquarefeet < 10_000]
df = df[df.calculatedfinishedsquarefeet > 400]

# remove tax percent of less than 1% and more than 100%
df = df[df.taxpercent > .0099]
df = df[df.taxpercent < 0.03]


In [ ]:
df.shape

In [ ]:
# Final df Records / Original df Records
2013640 / 2152863

In [ ]:
df.describe(include='all').T

## Maintain Columns

### Reorder Columns

In [ ]:
# List the columns
df.columns.values

#### Rearange the list manually

'bedroomcnt',  
'bathroomcnt',  
'bed_bath_ratio',  
'calculatedfinishedsquarefeet',  
'yearbuilt',  
'age',  
'taxvaluedollarcnt',  
'taxamount',  
'taxpercent',  
'fips',  
'name',  
'state'

In [ ]:
# Rearange Columns
df = df[['bedroomcnt', 
'bathroomcnt', 
'bed_bath_ratio', 
'calculatedfinishedsquarefeet',
'yearbuilt', 
'age',
'taxvaluedollarcnt', 
'taxamount', 
'taxpercent', 
'fips', 
'name', 
'state']]

In [ ]:
df.head().T

### Rename Columns

### Drop Columns

In [ ]:
# Drop Columns propertylandusetypeid and propertylandusedesc
# The index and discription were only needed to filter and verify
df = df.drop(columns=['propertylandusetypeid', 'propertylandusedesc']) 

### 3. Store all of the necessary functions to automate your process from acquiring the data to returning a cleaned dataframe with no missing values in your ```wrangle.py``` file. Name your final function ```wrangle_zillow```.

In [ ]:
# reset df to zero to test wrangle function
# delete csv file to fully test all wrangle functions
df = 0

In [ ]:
# use custom wrangle function to automate Acquire and Prepare
df = w.wrangle_zillow()

In [ ]:
# Check our work
df.T

In [ ]:
df.info()

In [ ]:
df.describe().T

### Personal Exploration

In [ ]:
# Custom function that returns nunique for each value in each column
ee.nunique_column_all(df)

In [ ]:
# Built filtered df's to explore curiousity
bathroomcnt_df = df[df['bathroomcnt'] > 18]
yearbuilt_df = df[df['yearbuilt'] < 1900]

In [ ]:
bathroomcnt_df.T

In [ ]:
yearbuilt_df.T

In [ ]:
# look at oldest property just out of curiosity
yearbuilt_df.min()